From 8945bebdec9157a25b4331676a24d004ca43feda Mon Sep 17 00:00:00 2001
From: Hendrik Brueckner <brueckner@redhat.com>
Date: Mon, 16 Jun 2014 12:55:59 -0400
Subject: [s390] sclp_early: Replace early_read_info_sccb with sccb_early

Message-id: <1402923362-2331-9-git-send-email-brueckner@redhat.com>
Patchwork-id: 84030
O-Subject: [RHEL7.1 PATCH 08/11] [s390] s390/sclp_early: Replace early_read_info_sccb with sccb_early
Bugzilla: 1107730
RH-Acked-by: Tony Camuso <tcamuso@redhat.com>
RH-Acked-by: Steve Best <sbest@redhat.com>

Description
-----------
commit 56e57a84a7856130f45009ce1713dc8ec8e59887
Author: Hendrik Brueckner <brueckner@linux.vnet.ibm.com>
Date:   Thu Dec 5 19:03:50 2013 +0100

    s390/sclp_early: Replace early_read_info_sccb with sccb_early

    Replace early_read_info_sccb and use sccb_early instead.
    Also saves some memory.

    Signed-off-by: Hendrik Brueckner <brueckner@linux.vnet.ibm.com>
    Reviewed-by: Michael Holzheu <holzheu@linux.vnet.ibm.com>
    Reviewed-by: Heiko Carstens <heiko.carstens@de.ibm.com>
    Signed-off-by: Martin Schwidefsky <schwidefsky@de.ibm.com>

Bugzilla
--------
BZ 1107730
https://bugzilla.redhat.com/show_bug.cgi?id=1107730

Upstream status of the patch
----------------------------
The patch is upstream as of kernel version
http://git.kernel.org/?p=linux/kernel/git/torvalds/linux-2.6.git;a=commitdiff;h=56e57a84a7856130f45009ce1713dc8ec8e59887

Brew
----
https://brewweb.devel.redhat.com/taskinfo?taskID=7572070

Test status
-----------
The patch has been tested and fixes the problem.
The fix has been verified by the IBM test department.

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/drivers/s390/char/sclp_early.c b/drivers/s390/char/sclp_early.c
index e4e5b32..dc4ed04 100644
--- a/drivers/s390/char/sclp_early.c
+++ b/drivers/s390/char/sclp_early.c
@@ -35,8 +35,7 @@ struct read_info_sccb {
  u8 _reserved5[4096 - 112]; /* 112-4095 */
 } __packed __aligned(PAGE_SIZE);
 
-static __initdata struct read_info_sccb early_read_info_sccb;
-static __initdata char sccb_early[PAGE_SIZE] __aligned(PAGE_SIZE);
+static char sccb_early[PAGE_SIZE] __aligned(PAGE_SIZE) __initdata;
 static unsigned long sclp_hsa_size;
 static struct sclp_ipl_info sclp_ipl_info;
 
@@ -63,14 +62,12 @@ out:
  return rc;
 }
 
-static int __init sclp_read_info_early(void)
+static int __init sclp_read_info_early(struct read_info_sccb *sccb)
 {
  int rc, i;
- struct read_info_sccb *sccb;
  sclp_cmdw_t commands[] = {SCLP_CMDW_READ_SCP_INFO_FORCED,
       SCLP_CMDW_READ_SCP_INFO};
 
- sccb = &early_read_info_sccb;
  for (i = 0; i < ARRAY_SIZE(commands); i++) {
   do {
    memset(sccb, 0, sizeof(*sccb));
@@ -92,12 +89,11 @@ static int __init sclp_read_info_early(void)
 
 static void __init sclp_facilities_detect(void)
 {
- struct read_info_sccb *sccb;
+ struct read_info_sccb *sccb = (void *) &sccb_early;
 
- if (sclp_read_info_early())
+ if (sclp_read_info_early(sccb))
   return;
 
- sccb = &early_read_info_sccb;
  sclp_facilities = sccb->facilities;
  sclp_fac84 = sccb->fac84;
  if (sccb->fac85 & 0x02)
-- 
1.7.1